### ISSUE TITLE
Firebase Connection Issue

### ISSUE THREAD

**Issue by TheBloop** [firebase/quickstart-unity#875]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2020.1.0b5.3485
  * Firebase Unity SDK version: latest
  * Source you installed the SDK: Unity Package Manager
  * Problematic Firebase Component: Database
  * Platform you are using the Unity editor on: Windows
  * Platform you are targeting: Android

### [REQUIRED] Please describe the issue here:

I'm new to Unity Firebase. I followed a tutorial on how to connect the Firebase to Unity and uploaded and imported all the necessary files (the json file and the database package). However I noticed that unlike in the tutorial, I don't have a "Firebase" option in the "Window" tab. I've been told that it doesn't matter so I just ignored it. I wrote a code to write to Unity and whenever I click start I have an error. 

**The code:**

```
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Firebase.Extensions;
using UnityEngine.Events;
using Firebase.Database;
using Firebase;
using Firebase.Unity.Editor;

public class Firebasetest : MonoBehaviour
{
    private FirebaseDatabase _database;
    public UnityEvent OnFirebaseInitialized = new UnityEvent();

    void Start()
    {
        Debug.Log("testing fb");
        FirebaseApp.CheckAndFixDependenciesAsync().ContinueWithOnMainThread(continuation: task =>
        {
            if (task.Exception != null)
            {
                Debug.LogError(message: $"failed to fb with {task.Exception}");
                return;
            }
            Debug.Log("connected fb");
            OnFirebaseInitialized.Invoke();
        });

        _database = FirebaseDatabase.DefaultInstance;
        _database.GetReference("Connection").SetValueAsync("Conncected!");
    }
}
```

**The error:**
` Cannot close. Connection has been closed.
UnityEngine.Debug:LogError(Object)
Firebase.Platform.FirebaseLogger:LogMessage(PlatformLogLevel, String) (at Z:/tmp/tmp.tpC7OdQUHa/firebase/app/client/unity/src/Unity/FirebaseLogger.cs:95)
Firebase.LogUtil:LogMessage(LogLevel, String) (at Z:/tmp/tmp.JYnsY9apCV/firebase/app/client/unity/proxy/LogUtil.cs:68)
Firebase.LogUtil:LogMessageFromCallback(LogLevel, String) (at Z:/tmp/tmp.JYnsY9apCV/firebase/app/client/unity/proxy/LogUtil.cs:76)
Firebase.AppUtilPINVOKE:PollCallbacks()
Firebase.AppUtil:PollCallbacks() (at Z:/tmp/tmp.JYnsY9apCV/firebase/app/client/unity/proxy/AppUtil.cs:32)
Firebase.Platform.FirebaseAppUtils:PollCallbacks() (at Z:/tmp/tmp.JYnsY9apCV/firebase/app/client/unity/proxy/FirebaseAppUtils.cs:33)
Firebase.Platform.FirebaseHandler:Update() (at Z:/tmp/tmp.tpC7OdQUHa/firebase/app/client/unity/src/Unity/FirebaseHandler.cs:208)
Firebase.Platform.FirebaseMonoBehaviour:Update() (at Z:/tmp/tmp.tpC7OdQUHa/firebase/app/client/unity/src/Unity/FirebaseMonoBehaviour.cs:45)`

#### Steps to reproduce:

Have you been able to reproduce this issue with just the Firebase Unity quickstarts (this GitHub project)?
What's the issue repro rate? (eg 100%, 1/5 etc)

What happened? How can we make the problem occur?
This could be a description, log/console output, etc.

If you have a downloadable sample project that reproduces the bug you're reporting, you will
likely receive a faster response on your issue.

**Comment by chkuang-g** [comment]
Hi @TheBloop 

Just to make sure by `latest` you mean `6.16.1`, since a bug related to realtime database when play in editor was fixed in [6.16.1](https://firebase.google.com/support/release-notes/unity#version_6161_-_october_20_2020).

Also, try to reference to Database instance AFTER `CheckAndFixDependenciesAsync()` is finished, for instance, inside that continuation function.  See the following code.

```
        FirebaseApp.CheckAndFixDependenciesAsync().ContinueWithOnMainThread(continuation: task =>
        {
            if (task.Exception != null)
            {
                Debug.LogError(message: $"failed to fb with {task.Exception}");
                return;
            }
            Debug.Log("connected fb");

            _database = FirebaseDatabase.DefaultInstance;
            _database.GetReference("Connection").SetValueAsync("Conncected!");

            OnFirebaseInitialized.Invoke();
        });

```

This [quickstart](https://github.com/firebase/quickstart-unity/blob/f997826d90f0a1b535880c3990ed94e619628217/database/testapp/Assets/Firebase/Sample/Database/UIHandler.cs#L57) should give you a pretty good idea how to use Firebase API.

Let me know if this helps!
Shawn

**Comment by TheBloop** [comment]
> Hi @TheBloop
> 
> Just to make sure by `latest` you mean `6.16.1`, since a bug related to realtime database when play in editor was fixed in [6.16.1](https://firebase.google.com/support/release-notes/unity#version_6161_-_october_20_2020).
> 
> Also, try to reference to Database instance AFTER `CheckAndFixDependenciesAsync()` is finished, for instance, inside that continuation function. See the following code.
> 
> ```
>         FirebaseApp.CheckAndFixDependenciesAsync().ContinueWithOnMainThread(continuation: task =>
>         {
>             if (task.Exception != null)
>             {
>                 Debug.LogError(message: $"failed to fb with {task.Exception}");
>                 return;
>             }
>             Debug.Log("connected fb");
> 
>             _database = FirebaseDatabase.DefaultInstance;
>             _database.GetReference("Connection").SetValueAsync("Conncected!");
> 
>             OnFirebaseInitialized.Invoke();
>         });
> ```
> 
> This [quickstart](https://github.com/firebase/quickstart-unity/blob/f997826d90f0a1b535880c3990ed94e619628217/database/testapp/Assets/Firebase/Sample/Database/UIHandler.cs#L57) should give you a pretty good idea how to use Firebase API.
> 
> Let me know if this helps!
> Shawn

Hi, thank you for your response! I copied the code into mine, still has the same connection issue. However, you did bring up the SDK. I'm not sure how to check what version of SDK my unity is using and I tried looking it up online. I downloaded some files that the Unity site provided to keep the SDK up to date, but I'm not sure how these packages relate to SDK. How can I check what version my SDK is?

**Comment by patm1987** [comment]
Hi @TheBloop , sorry for having let this sit stale for awhile.

First, to figure out which version of the Firebase SDK you have installed. You can look under `Assets/Firebase/Editor` for files that more or less look like `*_version-<version number>_manifest`. For example, in this screenshot, you can see that I have 7.0.0 of Auth and Database:
![image](https://user-images.githubusercontent.com/1111611/102251369-f0883280-3ec1-11eb-8e50-7c66481db42e.png)

If that doesn't help, you can also look for a file called `AppDependencies.xml` in the same directory. In there you'll see the version in `"Firebase/Core"` or `firebase-app-unity` entries:
```
<!-- Copyright (C) 2018 Google Inc. All Rights Reserved.

FirebaseApp iOS and Android Dependencies.
-->

<dependencies>
  <iosPods>
    <iosPod name="Firebase/Core" version="7.0.0" minTargetSdk="9.0">
    </iosPod>
  </iosPods>
  <androidPackages>
    <androidPackage spec="com.google.firebase:firebase-common:19.3.1">
    </androidPackage>
    <androidPackage spec="com.google.firebase:firebase-analytics:18.0.0">
    </androidPackage>
    <androidPackage spec="com.google.android.gms:play-services-base:17.5.0">
    </androidPackage>
    <androidPackage spec="com.google.firebase:firebase-app-unity:7.0.0">
      <repositories>
        <repository>Assets/Firebase/m2repository</repository>
      </repositories>
    </androidPackage>
  </androidPackages>
</dependencies>
```

There are a few things we can try at this point.

1) The latest version is `7.0.1`, which is newer than what you would have been running before. I would recommend trying to upgrade to this and seeing if the problem goes away. You will have to upgrade everything that's in your project (the Unity plugin won't work well with, say, Database at 7.0.1 and Auth at 6.16.1).

2) It does worry me that you don't have a "Firebase" option under Window. This might indicate that the SDK hasn't been properly imported or that you have some other build error (screenshot attached from MacOS):
![image](https://user-images.githubusercontent.com/1111611/102251955-b53a3380-3ec2-11eb-9630-a525a6433651.png)

Let me know if it comes back after importing a newer SDK.

3) I'm curious if you have some permissions issue (or you just don't have the Database setup). Make sure that you see _something_ when you click on "Realtime Database" in the Firebase console:
![image](https://user-images.githubusercontent.com/1111611/102252306-2b3e9a80-3ec3-11eb-90e7-fddf3279aa03.png)

And you can try making the rules (temporarily!) open just to verify:
```
{
  "rules": {
    ".read": true,
    ".write": true
  }
}
```

4) Sometimes you have to re-download your `google-services.json` file. Follow [the instructions here](https://support.google.com/firebase/answer/7015592), and make sure that you _only have one in your project_.

5) Download the [Database quickstart app](https://github.com/firebase/quickstart-unity/tree/master/database/testapp) and configure it to hit your backend (add your `google-services.json` and make sure you update your package manifest). Let us know if the issue persists like this.

6) Identifying a project in Firebase is a combination of a SHA-1 and Package ID. If another project matches your SHA-1 and Package ID, you may run into connection issues. Verify that you don't have any other similar projects and try changing your package id/registering another Android app.

7) It may be a local Firewall setting. Try switching your network (ex: disabling wifi and using cell), and verify that you can reach `*.firebaseio.com` (or the IP address reported by `https://<PROJECT_NAME>.firebaseio.com/.settings/owner.json` - whitelisting the IP address is _strongly discouraged_ as it may change without notice).

If none of these work, I think I'll have to see further logs to see if anything is up. And it would _definitely_ help to let us know what the result of step 5 is (testing in the quickstart).

Thanks,
--Patrick
